compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -527,94 +527,63 @@ LogicalResult getDmaCpyNdOpProducersAndConsumers(
-    splitDim = -1;
-  } else {
-    return op.emitOpError() << "failed to find a dimension for splitting";
-  }
-
-  assert(splitDim < memrefShape.size() &&
-         "the dimension to be split on should be smaller than the number of "
-         "dimensions in the shape");
-  splitFactor = memrefShape[splitDim];
-  if (ShapedType::isDynamic(splitFactor)) {
-    return op.emitOpError()
-           << "a dynamic size on the split dimension is not supported";
+using OffsetIndexAndNewOffsetT = std::tuple<std::optional<size_t>, int64_t>;
+
+/// Utility to return the index of the offsets array that refers to newly
+/// splitted objectFifo and the respective offset value. Note that there might
+/// not be a dimension with `stride == sizeAfterSplit`, in which case an offset
+/// index can't be returned and the correct offset is `0`.
+FailureOr<OffsetIndexAndNewOffsetT> getOffsetIndexAndOffset(
+    ArrayRef<OpFoldResult> offsets, ArrayRef<OpFoldResult> sizes,
+    ArrayRef<OpFoldResult> strides, size_t sizeAfterSplit,
+    function_ref<InFlightDiagnostic()> emitError) {
+  SmallVector<size_t> offsetIndices;
+  for (auto iter : llvm::enumerate(llvm::zip(strides, offsets))) {
+    std::optional<int64_t> maybeStride =
+        getConstantIntValue(std::get<0>(iter.value()));
+    std::optional<int64_t> maybeOffset =
+        getConstantIntValue(std::get<1>(iter.value()));
+    if (maybeStride.has_value() && maybeOffset.has_value() &&
+        maybeStride.value() == sizeAfterSplit && maybeOffset.value() != 0) {
+      offsetIndices.push_back(iter.index());
+    }
   }
-  return success();
-}
-
-/// Utility to get the split dimension and factor from a L3->L2 dma op.
-LogicalResult getSplitDimAndFactorFromDma(AMDAIE::DmaCpyNdOp op,
-                                          int64_t &splitDim,
-                                          int64_t &splitFactor,
-                                          int64_t &splitDimInL2Dma) {
-  if (!op->use_empty())
-    return op.emitOpError() << "can't be split because it has uses";
-
-  // Get the split dim from L2 objectFifo.
-  LogicalObjectFifoFromMemrefOp srcObjectFifo = op.getSourceObjectFifo();
-  LogicalObjectFifoFromMemrefOp tgtObjectFifo = op.getTargetObjectFifo();
-
-  LogicalObjectFifoFromMemrefOp l2ObjectFifo;
-  FailureOr<bool> l2DmaTransposed;
-  if (srcObjectFifo.getMemorySpaceAsUInt() == 1) {
-    l2ObjectFifo = srcObjectFifo;
-    l2DmaTransposed = isDmaTransposedOnSourceSide(op);
-  } else if (tgtObjectFifo.getMemorySpaceAsUInt() == 1) {
-    l2ObjectFifo = tgtObjectFifo;
-    l2DmaTransposed = isDmaTransposedOnTargetSide(op);
-  } else {
-    return op.emitOpError()
-           << "the input dma should have source or target in L2 memory space";
+  if (offsetIndices.size() > 1)
+    return emitError() << "multiple offset indices found";
+  int64_t size{1};
+  int64_t offset{0};
+  std::optional<size_t> maybeOffsetIdx;
+  if (offsetIndices.size() == 1) {
```

Comment:
I think you can remove this condition if you have a separate check for offsetIndices.empty()?